src/scratch_impl.h

```diff
@@ -62,21 +77,34 @@ static int secp256k1_scratch_allocate_frame(secp256k1_scratch* scratch, size_t n
     }
 }
 
-static void secp256k1_scratch_deallocate_frame(secp256k1_scratch* scratch) {
+static void secp256k1_scratch_deallocate_frame(const secp256k1_callback* error_callback, secp256k1_scratch* scratch) {
+    if (memcmp(scratch->magic, "scratch", 8) != 0) {
+        secp256k1_callback_call(error_callback, "invalid scratch space");
+        return;
+    }
+
     VERIFY_CHECK(scratch->frame > 0);
-    scratch->frame -= 1;
-    free(scratch->data[scratch->frame]);
+
+    scratch->frame--;
+    scratch->data = (void *) ((char *) scratch->data - scratch->frame_size[scratch->frame]);
 }
 
-static void *secp256k1_scratch_alloc(secp256k1_scratch* scratch, size_t size) {
+static void *secp256k1_scratch_alloc(const secp256k1_callback* error_callback, secp256k1_scratch* scratch, size_t size) {
     void *ret;
-    size_t frame = scratch->frame - 1;
+    size_t frame;
+
+    if (memcmp(scratch->magic, "scratch", 8) != 0) {
+        secp256k1_callback_call(error_callback, "invalid scratch space");
+        return NULL;
+    }
+
+    frame = scratch->frame - 1;
     size = ((size + ALIGNMENT - 1) / ALIGNMENT) * ALIGNMENT;
 
     if (scratch->frame == 0 || size + scratch->offset[frame] > scratch->frame_size[frame]) {
```

Comment:
fwiw, this size check can be circumvented by setting size such that `size + scratch->offset[frame]` overflows

Reply:
Fixed with checkpointing